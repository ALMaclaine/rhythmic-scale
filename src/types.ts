export interface RhythmicScaleParams {
    baseFont?: number;
    baseBrowserFontSize?: number;
    lineHeightFactor?: number;
}
