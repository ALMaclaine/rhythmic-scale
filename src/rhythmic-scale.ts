import {RhythmicScaleParams} from "./types";
import {MusicalRatios, ratioToPower} from "musical-ratios";
import {ensureNotPxRem} from "empxrem";

const rhythmicScaleDefaultParams: RhythmicScaleParams = {
    baseFont: 10,
    baseBrowserFontSize: 16,
    lineHeightFactor: 1.2
}

function scaledSize(ratio: MusicalRatios | number, level: number, size: number = 1) {
    return size * ratioToPower(ratio, level);
}

export class RhythmicScale {
     public baseFont: number;
     public baseBrowserFontSize: number;
     public lineHeightFactor: number;

    constructor(params: RhythmicScaleParams = {}) {
        const {baseFont, baseBrowserFontSize, lineHeightFactor} = {...rhythmicScaleDefaultParams, ...params};
        this.baseFont = baseFont;
        this.baseBrowserFontSize = baseBrowserFontSize;
        this.lineHeightFactor = lineHeightFactor;
    }

    scaledStyles(ratio: MusicalRatios | number, level: number, min?: number): {lineHeight: string, size: string} {
        const scaled = scaledSize(ratio, level, this.baseFont);
        return {
            size: ensureNotPxRem(scaled, this.baseBrowserFontSize),
            lineHeight: ensureNotPxRem(scaled * this.lineHeightFactor, this.baseBrowserFontSize, min)
        }
    }
}
