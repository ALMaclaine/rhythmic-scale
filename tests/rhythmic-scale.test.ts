import * as rhythmicScale from '../src';

describe('rhythmic-scale', () => {
    it('Has Correct API', () => {
        const keys = Object.keys(rhythmicScale);
        expect(keys.length).toBe(1);
        expect(keys).toMatchSnapshot();
    });
});
