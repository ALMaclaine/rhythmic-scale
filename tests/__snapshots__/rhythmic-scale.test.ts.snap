// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rhythmic-scale Has Correct API 1`] = `
Array [
  "RhythmicScale",
]
`;
